### ISSUE TITLE
Disallow {[string]} array notation

### ISSUE THREAD

**Issue by yvele** [gajus/eslint-plugin-jsdoc#495]
I know that the `{[string]}` array notation doesn't exists at all.. but I can find it quite often in my company code base, and I would like to restrict it (ideally with an auto fix). Does that make sense?

Is it possible to disallow and fix this kind of array notation?

❌ This should be disallowed 

```js
/**
 * @param {[string]} foo
 */
```

And fixed depending on `preferredTypes` to either

```js
/**
 * @param {Array<string>} foo
 */
```

```js
/**
 * @param {Array.<string>} foo
 */
```

```js
/**
 * @param {string[]} foo
 */
```

**Comment by brettz9** [comment]
`[string]` is a valid type for tuples in TypeScript: https://www.typescriptlang.org/docs/handbook/basic-types.html#tuple .

Both [jsdoc](https://github.com/jsdoc/jsdoc/issues/1703) and [Closure](https://github.com/google/closure-compiler/issues/379) have open issues to add this support. (There are also issues filed with [closure](https://github.com/google/closure-compiler/issues/1970) and [jsdoc](https://github.com/jsdoc/jsdoc/issues/1693) which may also be of interest in seeking to add a way for jsdoc to add names for the Tuple components; but these are not supported in jsdoctypeparser.)

There is [a jsdoctypeparser issue](https://github.com/jsdoctypeparser/jsdoctypeparser/issues/96) filed, but the [PR](https://github.com/jsdoctypeparser/jsdoctypeparser/pull/97) I put together for it needs research toward which rule differences exist for which modes, and I haven't found time for this research. However, I should be able to complete the PR earlier and just add to the different modes gradually, such as treating tuples as disallowed for `"settings":{"jsdoc":{"mode": "jsdoc"}}` and `"settings":{"jsdoc":{"mode": "closure"}}` (and tying it to a `settings.jsdoc.modeVersion` or `mode: ["jsdoc", <version>]` as well in case future versions of jsdoc/Closure do support this). But still would need time to get to it (or someone to revise the PR).

So that aspect is really a jsdoctypeparser, but as with your other issue recently filed, the fixer does have something for us to fix on our end.

**Comment by yvele** [comment]
Thanks for the super explicit and referenced clarification! (I like your style 👍 )

Yeah, I would like to disable the tuple type in the projects that doesn't use TypeScript

**Comment by brettz9** [comment]
Closing as this is effectively a dupe of part of #356, though I have made explicit mention there of this use case.

**Comment by yvele** [comment]
Hum FYI I just remembered where I saw the `{[type]}` array notation: https://github.com/jsdoc/jsdoc/issues/1073#issuecomment-312524697

```js
/**
 * Foo function.
 *
 * @param {[number, string]} bar - The bar array.
 */
function foo(bar) {}
```

**Comment by yvele** [comment]
I've also seen a mixed `{[type]}` array notation here https://stackoverflow.com/a/46517184/1480391 (back in 2017)

```js
/** @type {[Foo, Bar]} */
const arr = [new Foo(), new Bar()];
```
